apm-agent-plugin-sdk/src/main/java/co/elastic/apm/agent/sdk/logging/LoggerFactory.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.sdk.logging;
+
+public class LoggerFactory {
+
+    private static ILoggerFactory iLoggerFactory;
```

Comment:
How about either making this final or volatile?

Reply:
We could use a similar pattern as in `WeakConcurrent` or `DynamicTransformer` to make the field final.

https://github.com/elastic/apm-agent-java/blob/46e81e53670afb42c2e5ae57ae0849ab7d50b2ca/apm-agent-plugin-sdk/src/main/java/co/elastic/apm/agent/sdk/weakconcurrent/WeakConcurrent.java#L26-L35

Reply:
> How about either making this final or volatile?

Good catch! I could have sworn I made it `volatile`...

> We could use a similar pattern as in WeakConcurrent or DynamicTransformer to make the field final.

No, that won't work well here. I want to start using our log4j factory bridge only after [setting our log4j configuration factory](https://github.com/elastic/apm-agent-java/pull/2390/files#diff-2f5c514f954a475ff220cebc3a8551dce7915534f14d15bcffa86021ad108403R326), which relies on the list of `ConfigurationSource`s. We can change our entire initialization sequence for this purpose, but that doesn't seems justified.